File: software/examples/c/example_callback.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_dual_button.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change XYZ to the UID of your Dual Button Bricklet

// Callback function for state changed callback
void cb_state_changed(uint8_t button_l, uint8_t button_r, uint8_t led_l, uint8_t led_r,
                      void *user_data) {
	(void)led_l; // avoid unused parameter warning
	(void)led_r; // avoid unused parameter warning
	(void)user_data; // avoid unused parameter warning

	if(button_l == DUAL_BUTTON_BUTTON_STATE_PRESSED) {
		printf("Left button pressed\n");
	} else {
		printf("Left button released\n");
	}

	if(button_r == DUAL_BUTTON_BUTTON_STATE_PRESSED) {
		printf("Right button pressed\n");
	} else {
		printf("Right button released\n");
	}

	printf("\n");
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	DualButton db;
	dual_button_create(&db, UID, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		return 1;
	}
	// Don't use device before ipcon is connected

	// Register state changed callback to function cb_state_changed
	dual_button_register_callback(&db,
	                              DUAL_BUTTON_CALLBACK_STATE_CHANGED,
	                              (void *)cb_state_changed,
	                              NULL);

	printf("Press key to exit\n");
	getchar();
	dual_button_destroy(&db);
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
	return 0;
}

File: software/examples/php/ExampleCallback.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletDualButton.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletDualButton;

const HOST = 'localhost';
const PORT = 4223;
const UID = 'XYZ'; // Change XYZ to the UID of your Dual Button Bricklet

// Callback function for state changed callback
function cb_stateChanged($button_l, $button_r, $led_l, $led_r)
{
    if ($button_l == BrickletDualButton::BUTTON_STATE_PRESSED) {
        echo "Left Button: Pressed\n";
    } elseif ($button_l == BrickletDualButton::BUTTON_STATE_RELEASED) {
        echo "Left Button: Released\n";
    }

    if ($button_r == BrickletDualButton::BUTTON_STATE_PRESSED) {
        echo "Right Button: Pressed\n";
    } elseif ($button_r == BrickletDualButton::BUTTON_STATE_RELEASED) {
        echo "Right Button: Released\n";
    }

    echo "\n";
}

$ipcon = new IPConnection(); // Create IP connection
$db = new BrickletDualButton(UID, $ipcon); // Create device object

$ipcon->connect(HOST, PORT); // Connect to brickd
// Don't use device before ipcon is connected

// Register state changed callback to function cb_stateChanged
$db->registerCallback(BrickletDualButton::CALLBACK_STATE_CHANGED, 'cb_stateChanged');

echo "Press ctrl+c to exit\n";
$ipcon->dispatchCallbacks(-1); // Dispatch callbacks forever

?>
